### PR TITLE
Bug 1314270: force dc reconcilation on canceled deployments

### PR DESCRIPTION
**Pull request by kargakis**
Force a deploymentconfig reconcilation when its running deployment
is canceled instead of relying on the deploymentconfig cache sync
interval for rolling back.

Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1314270

@ironcladlou @smarterclayton PTAL

### CODE REVIEW

**Comment by kargakis** [comment]
The only other way to handle https://bugzilla.redhat.com/show_bug.cgi?id=1314270 I can think of that doesn't include any changes to the automatic rollback process is to force a reconcile of a deploymentconfig (by adding an annotation most probably) in the deployer controller (right after a cancelled deployment is marked as failed).

**Comment by smarterclayton** [comment]
We can't shorten this - it would relist all dcs every 30 seconds (unless that has changed?).  if we aren't listing against etcd in the resync then this is possibly ok, but I don't see how this won't dramatically affect server performance in contented scenarios

**Comment by kargakis** [comment]
Hasn't changed to my knowledge. So if we are not going to change the interval the only solution left for https://bugzilla.redhat.com/show_bug.cgi?id=1314270 is to trigger a dc reconcilation from the deployer controller once the deployment is marked as failed. This is even better than just decreasing the interval because the dc will be reconciled and the latest complete deployment will be rolled back up on the spot. 

My only concern is update conflicts - we are creating yet another update site for deploymentconfigs. Is this a valid concern? I know we can manually retry updates but it is tedious to pick those you desperately need to happen and have them succeed. Having something like https://github.com/kubernetes/kubernetes/issues/21479 would be helpful.

**Comment by smarterclayton** [comment]
Hrm - I think we can live with it, since you're just trying to apply a
mutation after the fact (a few retries is sufficient).  We do have the
RetryOnConflict helper which should make this cleaner than it was before.

On Tue, Mar 22, 2016 at 10:02 AM, Michail Kargakis <notifications@github.com

**Comment by kargakis** [comment]
Updated to force dc reconcilation on canceled deployments. We can argue whether the deployer pod controller or the deployment controller is the right place for the dc update. I went with the former since the update will happen as soon as the deployment transitions to canceled->failed.

cc: @ironcladlou

**Comment by kargakis** [comment]
[test]

**Comment by ironcladlou** [comment]
I think this approach makes sense. I'm not concerned about multiple update sites in this instance because we're making an update to the status that doesn't affect version or scaling behavior.

The only other way I thought of off the top of my head would be to consolidate the controllers into a single controller and switch to a pattern which would allow the RC handling part to explicitly enqueue the DC for the DC handler to pick up rather than mutating the DC to coerce queueing in a separate controller. Obviously that's a much larger discussion...

**Comment by kargakis** [comment]
> The only other way I thought of off the top of my head would be to consolidate the controllers into a single controller and switch to a pattern which would allow the RC handling part to explicitly enqueue the DC for the DC handler to pick up rather than mutating the DC to coerce queueing in a separate controller. Obviously that's a much larger discussion...

Yea, explicitly enqueing the dc would be optimal.

**Comment by smarterclayton** [comment]
If we need to unify controllers, it probably means the controller
responsibilities aren't well factored.  The hard part with cancel is that
it has to be recorded transactionally on something that corresponds to a
thing that needs to be cancelled.  As long as we have a deployment process
(in the pod) and a need to transactionally save whether a deployment has
been started (on the RC) we'll always need to make the record of
cancellation on the RC.

On Wed, Mar 23, 2016 at 10:34 AM, Michail Kargakis <notifications@github.com

**Comment by kargakis** [comment]
Updated to use RetryOnConflict for the dc update. Any other comments here?

File: pkg/deploy/controller/deployerpod/controller.go [hunk 1]
```diff
@@ -114,42 +127,33 @@ func (c *DeployerPodController) Handle(pod *kapi.Pod) error {
...
+			// If we are unable to get the deployment config, then the deploymentconfig controller will
+			// perform its duties once the resync interval forces the deploymentconfig to be reconciled.
+			name := deployutil.DeploymentConfigNameFor(deployment)
+			if config, err := c.client.DeploymentConfigs(deployment.Namespace).Get(name); err == nil {
+				if config.Annotations == nil {
+					config.Annotations = make(map[string]string)
+				}
+				config.Annotations[deployapi.DeploymentCancelledAnnotation] = strconv.Itoa(config.Status.LatestVersion)
+				return kclient.RetryOnConflict(kclient.DefaultRetry, func() error {
+					_, err := c.client.DeploymentConfigs(config.Namespace).Update(config)
```
Comment (ironcladlou) [inline pkg/deploy/controller/deployerpod/controller.go]: Do you need to re-fetch the config on conflict before retrying an update?

Comment (kargakis) [inline pkg/deploy/controller/deployerpod/controller.go]: Yup, good catch. That was the reason I fetched the config in the first place and didn't use the decoded one but it seems I missed this:)

File: pkg/deploy/controller/deployerpod/controller.go [hunk 2]
```diff
@@ -97,14 +97,8 @@ func (c *DeployerPodController) Handle(pod *kapi.Pod) error {
 			nextStatus = deployapi.DeploymentStatusRunning
 		}
 	case kapi.PodSucceeded:
-		// Detect failure based on the container state
 		nextStatus = deployapi.DeploymentStatusComplete
-		for _, info := range pod.Status.ContainerStatuses {
```
Comment (kargakis) [inline pkg/deploy/controller/deployerpod/controller.go]: @ironcladlou @smarterclayton I am removing this as [kubelet](https://github.com/kubernetes/kubernetes/blob/f89432e44b7aeac6e48d16fc34952e644a388b8d/pkg/kubelet/kubelet.go#L3230) already does that for us.

File: pkg/deploy/controller/deployerpod/controller.go [hunk 3]
```diff
@@ -142,16 +136,18 @@ func (c *DeployerPodController) Handle(pod *kapi.Pod) error {
 			// If we are unable to get the deployment config, then the deploymentconfig controller will
 			// perform its duties once the resync interval forces the deploymentconfig to be reconciled.
 			name := deployutil.DeploymentConfigNameFor(deployment)
-			if config, err := c.client.DeploymentConfigs(deployment.Namespace).Get(name); err == nil {
+			kclient.RetryOnConflict(kclient.DefaultRetry, func() error {
```
Comment (kargakis) [inline pkg/deploy/controller/deployerpod/controller.go]: @ironcladlou updated to re-fetch. Note that I am deliberately ignoring any error here (other than conflicts) as the dc controller will eventually resync the config.

**Comment by ironcladlou** [comment]
LGTM now.

**Comment by kargakis** [comment]
@smarterclayton any objections to having this for 1.2? First commit is a honest refactoring, and I perceive the second as more than a nice-to-have, since depending on when the deployment was marked as canceled, it can take up to 2m (the resync interval of the dc controller) for it to be scaled down.

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
Sgtm, Lgtm [merge]

**Pull request merged**